test/FunctionalTests/Microsoft.OData.Client.Tests/Serialization/MockDelayedHttpClientHandler.cs

```diff
@@ -0,0 +1,82 @@
+﻿//---------------------------------------------------------------------
+// <copyright file="MockDelayedHttpClientHandler.cs" company="Microsoft">
+//      Copyright (C) Microsoft Corporation. All rights reserved. See License.txt in the project root for license information.
+// </copyright>
+//---------------------------------------------------------------------
+
+using System;
+using System.Diagnostics;
+using System.Net.Http;
+using System.Threading;
+
+namespace Microsoft.OData.Client.Tests.Serialization
+{
+    /// <summary>
+    /// A mock HttpClientHandler that responds to HTTP requests with a mocked
+    /// response after a specified time delay.
+    /// </summary>
+    internal sealed class MockDelayedHttpClientHandler : MockHttpClientHandler
+    {
+        private readonly int _delay;
+        public MockDelayedHttpClientHandler(string expectedResponse, int delayMilliseconds) : base(expectedResponse)
+        {
+            _delay = delayMilliseconds;
+        }
+
+        /// <summary>
+        /// Callback that is triggered when a request has started.
+        /// This can be used to coordinate with a task on a separate
+        /// thread that should run only after a request has started.
+        /// For example, we may want to test aborting a request after
+        /// the request has started.
+        /// </summary>
+        public Action<HttpRequestMessage> OnRequestStarted { get; set; }
+
+#if NETCOREAPP
+        protected override HttpResponseMessage Send(HttpRequestMessage request, CancellationToken cancellationToken)
+        {
+            NotifyRequestStart(request);
+            Stopwatch stopwatch = new Stopwatch();
+            stopwatch.Start();
+
+            while (true)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                if (stopwatch.ElapsedMilliseconds > _delay)
+                {
+                    break;
+                }
+            }
+
+            stopwatch.Stop();
+            return base.Send(request, cancellationToken);
+        }
+#else
+        public override HttpResponseMessage Send(HttpRequestMessage request, CancellationToken cancellationToken)
+        {
+            NotifyRequestStart(request);
+            Stopwatch stopwatch = new Stopwatch();
+            stopwatch.Start();
+
+            while (true)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                if (stopwatch.ElapsedMilliseconds > _delay)
+                {
+                    break;
+                }
+            }
+
+            stopwatch.Stop();
+            return base.Send(request, cancellationToken);
+        }
+#endif

```

Comment:
The difference is that in `NETCOREAPP` we use `protected override` otherwise we use `public override`. Note that this class inherits from the test `MockHttpClientHandler` which in turns inherits from the framework's `HttpClientHandler`. In .NET, `HttpClientHandler` defines a `protected virtual` `Send` method, but that's not the case in .NET FX. So in NET Fx, the `MockHttpClientHandler` implements a `public virtual Send` overload, which is what is being overridden here in the `#else` block.

I could simplify this code by having the `#if-#else` block only on the method signature.

Reply:
> I could simplify this code by having the #if-#else block only on the method signature.

@habbes This makes more sense. We have done it that way in other places

Reply:
I made that change.